lxd/instance/drivers/driver_qemu_templates.go

```diff
@@ -538,190 +537,267 @@ func qemuDriveFirmwareSections(opts *qemuDriveFirmwareOpts) []cfgSection {
+		path:     opts.path,
+		sockFd:   fmt.Sprintf("%d", opts.proxyFD),
+	})
+}
+
+type qemuPCIPhysicalOpts struct {
+	dev         qemuDevOpts
+	devName     string
+	pciSlotName string
+	bootIndex   int
+}
+
+func qemuPCIPhysicalSections(opts *qemuPCIPhysicalOpts) []cfgSection {
+	deviceOpts := qemuDevEntriesOpts{
+		dev:     opts.dev,
+		pciName: "vfio-pci",
+		ccwName: "vfio-ccw",
+	}
+
+	entries := append(qemuDeviceEntries(&deviceOpts), []cfgEntry{
+		{key: "host", value: opts.pciSlotName},
+		{key: "bootIndex", value: fmt.Sprintf("%d", opts.bootIndex)},
+	}...)
+
+	return []cfgSection{{
+		// Devices use "lxd_" prefix indicating that this is a user named device.
+		name:    fmt.Sprintf(`device "dev-lxd_%s"`, opts.devName),
+		comment: fmt.Sprintf(`PCI card ("%s" device)`, opts.devName),
+		entries: entries,
+	}}
+}
+
+type qemuGPUDevPhysicalOpts struct {
+	dev         qemuDevOpts
+	devName     string
+	pciSlotName string
+	vgpu        string
+	vga         bool
+}
+
+func qemuGPUDevPhysicalSections(opts *qemuGPUDevPhysicalOpts) []cfgSection {
+	deviceOpts := qemuDevEntriesOpts{
+		dev:     opts.dev,
+		pciName: "vfio-pci",
+		ccwName: "vfio-ccw",
+	}
+
+	entries := qemuDeviceEntries(&deviceOpts)
+
+	if opts.vgpu != "" {
+		sysfsdev := fmt.Sprintf("/sys/bus/mdev/devices/%s", opts.vgpu)
+		entries = append(entries, cfgEntry{key: "sysfsdev", value: sysfsdev})
+	} else {
+		entries = append(entries, cfgEntry{key: "host", value: opts.pciSlotName})
+	}
+
+	if opts.vga {
+		entries = append(entries, cfgEntry{key: "x-vga", value: "on"})
+	}
+
+	return []cfgSection{{
+		// Devices use "lxd_" prefix indicating that this is a user named device.
+		name:    fmt.Sprintf(`device "dev-lxd_%s"`, opts.devName),
+		comment: fmt.Sprintf(`GPU card ("%s" device)`, opts.devName),
+		entries: entries,
+	}}
+}
+
+type qemuUSBOpts struct {
+	devBus        string
+	devAddr       string
+	multifunction bool
+	ports         int
+}
+
+func qemuUSBSections(opts *qemuUSBOpts) []cfgSection {
+	deviceOpts := qemuDevEntriesOpts{
+		dev: qemuDevOpts{
+			busName:       "pci",
+			devAddr:       opts.devAddr,
+			devBus:        opts.devBus,
+			multifunction: opts.multifunction,
+		},
+		pciName: "qemu-xhci",
+	}
+
+	sections := []cfgSection{{
+		name:    `device "qemu_usb"`,
+		comment: "USB controller",
+		entries: append(qemuDeviceEntries(&deviceOpts), []cfgEntry{
+			{key: "p2", value: fmt.Sprintf("%d", opts.ports)},
+			{key: "p3", value: fmt.Sprintf("%d", opts.ports)},
+		}...),
+	}}
+
+	for i := 1; i <= 3; i++ {
+		chardev := fmt.Sprintf("qemu_spice-usb-chardev%d", i)
+		sections = append(sections, []cfgSection{{
+			name: fmt.Sprintf(`chardev "%s"`, chardev),
+			entries: []cfgEntry{
+				{key: "backend", value: "spicevmc"},
+				{key: "name", value: "usbredir"},
+			},
+		}, {
+			name: fmt.Sprintf(fmt.Sprintf(`device "qemu_spice-usb%d"`, i)),
+			entries: []cfgEntry{
+				{key: "driver", value: "usb-redir"},
+				{key: "chardev", value: chardev},
+			},
+		}}...)
```

Comment:
Note that I've reordered chardev/device USB sections to make it possible creating in a single loop. If this is a problem I can keep the original order of sections by using 2 loops.